src/RiotPls.DataDragon/DataDragonClient.cs

```diff
@@ -0,0 +1,282 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Net.Http;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using RiotPls.DataDragon.Entities;
+using RiotPls.DataDragon.Enums;
+using RiotPls.DataDragon.Extensions;
+using RiotPls.DataDragon.Helpers;
+
+namespace RiotPls.DataDragon
+{
+    public sealed partial class DataDragonClient : IDataDragonClient
+    {
+        public const string Host = "https://ddragon.leagueoflegends.com";
+        public const string Api = "/api";
+        public const string Cdn = "/cdn";
+
+        public Language DefaultLanguage => _options.DefaultLanguage;
+
+        internal static DataDragonClient Instance { get; } = new DataDragonClient(DataDragonClientOptions.Default);
+
+        private readonly DataDragonClientOptions _options;
+        private readonly HttpClient _client;
+        private readonly JsonSerializerOptions _jsonSerializerOptions;
+        private readonly SemaphoreSlim _semaphore;
+        private readonly object _lock;
+        private ImmutableArray<GameVersion> _versions;
+        private ImmutableArray<Language> _languages;
+        private bool _isDisposed;
+
+        public DataDragonClient() : this(DataDragonClientOptions.Default)
+        {
+        }
+
+        public DataDragonClient(DataDragonClientOptions options)
+        {
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _client = new HttpClient
+            {
+                BaseAddress = new Uri(Host)
+            };
+            _jsonSerializerOptions = new JsonSerializerOptions
+            {
+                PropertyNameCaseInsensitive = true
+            };
+            _lock = new object();
+            _semaphore = new SemaphoreSlim(1, 1);
+        }
+
+        ~DataDragonClient()
+        {
+            Dispose();
+        }
+
+        /// <inheritdoc/>
+        public ValueTask<IReadOnlyList<GameVersion>> GetVersionsAsync()
```

Comment:
Change it to this:

```cs
        /// <inheritdoc/>
        public ValueTask<IReadOnlyList<Language>> GetLanguagesAsync()
        {
            ThrowIfDisposed();

            lock (_lock)
            {
                if (_options.CacheMode != CacheMode.None && !_languages.IsDefaultOrEmpty)
                    return new ValueTask<IReadOnlyList<Language>>(_languages);

                return new ValueTask<IReadOnlyList<Language>>(FetchLanguagesAsync());
            }
        }
```

Reply:
Oh sorry, this is for languages, well, do it for languages and version, it's pretty much the same code